sphinx/ext/autodoc/__init__.py

```diff
@@ -60,7 +60,7 @@
 py_ext_sig_re = re.compile(
     r'''^ ([\w.]+::)?            # explicit module name
           ([\w.]+\.)?            # module and/or class name(s)
-          (\w+)  \s*             # thing name
+          ([^.()]+)  \s*         # thing name
```

Comment:
I'd prefer to be more conservative here as you suggest - perhaps just add '-', ' ', ';' or etc to begin with.